src/mlpack/methods/neighbor_search/knn_low_resource_main.cpp

```diff
@@ -0,0 +1,188 @@
+  "and may be slow). You may specify a separate set of "
+  "reference points and query points, or just a reference set which will be "
+  "used as both the reference and query set."
+  "\n\n"
+  "For example, the following command will calculate the 5 nearest neighbors "
+  "of each point in " +
+    PRINT_DATASET("input") +
+    " and store the distances "
+    "in " +
+    PRINT_DATASET("distances") + " and the neighbors in " +
+    PRINT_DATASET("neighbors") +
+    ": "
+    "\n\n" +
+    PRINT_CALL("knn",
+               "k",
+               5,
+               "reference",
+               "input",
+               "neighbors",
+               "neighbors",
+               "distances",
+               "distances") +
+    "\n\n"
+    "The output is organized such that row i and column j in the neighbors "
+    "output matrix corresponds to the index of the point in the reference set "
+    "which is the j'th nearest neighbor from the point in the query set with "
+    "index i.  Row j and column i in the distances output matrix corresponds to"
+    " the distance between those two points.",
+  SEE_ALSO("NeighborSearch tutorial (k-nearest-neighbors)",
+           "@doxygen/nstutorial.html"),
+  SEE_ALSO("Tree-independent dual-tree algorithms (pdf)",
+           "http://proceedings.mlr.press/v28/curtin13.pdf"),
+  SEE_ALSO("mlpack::neighbor::NeighborSearch C++ class documentation",
+           "@doxygen/classmlpack_1_1neighbor_1_1NeighborSearch.html"));
+
+ // Define our input parameters that this program will take.
+ PARAM_MATRIX_OUT("distances", "Matrix to output distances into.", "d");
+ PARAM_UMATRIX_OUT("neighbors", "Matrix to output neighbors into.", "n");
+
+ // // The option exists to load or save models.
+//PARAM_MODEL_IN(KNNModel, "input_model", "Pre-trained kNN model.", "m");
+
+ // The user may specify a query file of query points and a number of nearest
+ // neighbors to search for.
+ PARAM_MATRIX_IN("query", "Matrix containing query points (optional).", "q");
+ PARAM_INT_IN("k", "Number of nearest neighbors to find.", "k", 0);
+
+static void
+mlpackMain()
+{
+  ReportIgnoredParam({ { "input_model", true } }, "tree_type");
+  ReportIgnoredParam({ { "input_model", true } }, "random_basis");
+
+  // The user should give something to do...
+  RequireAtLeastOnePassed(
+    { "k", "output_model" }, false, "no results will be saved");
+
+  // If the user specifies k but no output files, they should be warned.
+  if (CMD::HasParam("k")) {
+    RequireAtLeastOnePassed(
+      { "neighbors", "distances" },
+      false,
+      "nearest neighbor search results will not be saved");
+  }
+
+  // If the user specifies output files but no k, they should be warned.
+  ReportIgnoredParam({ { "k", false } }, "neighbors");
+  ReportIgnoredParam({ { "k", false } }, "distances");
+  ReportIgnoredParam({ { "k", false } }, "query");
+  KNNModel* knn;
+
+  // Load the model from file.
+ knn = CMD::GetParam<KNNModel*>("input_model");
```

Comment:
If you can remove this line (and the call to `Log::Info`) below, I think it should be possible to compile this entire program without any cereal support, which could reduce the size a bit more if you wanted to.  That said, it seems like cereal is pretty small, so maybe that support wouldn't be too expensive. :)